File: com_nvidia_gxf/gxf/core/logger.py
'''
 SPDX-FileCopyrightText: Copyright (c) 2023-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 SPDX-License-Identifier: Apache-2.0

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
'''
""" Logger
"""

import sys
import logging
from logging.handlers import RotatingFileHandler
import tempfile
from typing import Optional

CORE_PACKAGE_NAME: str = 'Core'


class StdOutFormatter(logging.Formatter):

    grey = "\x1b[38;5;248m"
    white = "\x1b[38;97m"
    yellow = "\x1b[33;20m"
    red = "\x1b[31;20m"
    bold_red = "\x1b[31;1m"
    reset = "\x1b[0m"
    format_ = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"

    FORMATS = {
        logging.DEBUG: grey + format_ + reset,
        logging.INFO: white + format_ + reset,
        logging.WARNING: yellow + format_ + reset,
        logging.ERROR: red + format_ + reset,
        logging.CRITICAL: bold_red + format_ + reset
    }

    def format(self, record):
        log_fmt = self.FORMATS.get(record.levelno)
        formatter = logging.Formatter(log_fmt)
        return formatter.format(record)

class CoreLogging:
    """ Class CoreLogger
    """
    _inited: bool = False

    @classmethod
    def is_inited(cls) -> bool:
        return cls._inited

    @classmethod
    def init_logger(
        cls) -> None:
        logger = cls.get_logger()
        logger.setLevel(logging.DEBUG)
        cls.clear_handlers(logger)

        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        stream_handler = logging.StreamHandler(stream=sys.stdout)
        stream_handler.setFormatter(StdOutFormatter())
        stream_handler.setLevel(logging.DEBUG) # change back
        logger.addHandler(stream_handler)
        fh = RotatingFileHandler(
            tempfile.gettempdir() + "//gxf_core.log", mode='a', maxBytes=1048576, backupCount=5)
        fh.setLevel(logging.DEBUG)
        fh.setFormatter(formatter)
        logger.addHandler(fh)

        cls._inited = True

    @classmethod
    def clear_handlers(cls, logger):
        handlers = logger.handlers.copy()
        for h in handlers:
            try:
                h.acquire()
                h.flush()
                h.close()
            except (OSError, ValueError):
                pass
            finally:
                h.release()
                logger.removeHandler(h)
        logger.handlers.clear()

    @classmethod
    def get_logger(cls, name: Optional[str] = None) -> logging.Logger:
        """ Get a Core component logger

        Parameters:
        name: a module name

        Returns: A Logger Instance
        """
        propagate = True
        name = CORE_PACKAGE_NAME
        if not name:
            propagate = False


        logger = logging.getLogger(name)
        logger.propagate = propagate
        return logger


get_logger = CoreLogging.get_logger
init_logger = CoreLogging.init_logger
is_inited = CoreLogging.is_inited

File: com_nvidia_gxf/gxf/logger/logger.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "gxf/logger/logger.hpp"

#include <memory>

namespace nvidia {

/// Namespace for the NVIDIA logger functionality.
namespace logger {

constexpr int kMaxSeverity = 255;

Logger::Logger(const std::shared_ptr<ILogger>& logger) : logger_(logger) {}

Logger::Logger(const LogFunction& func) : func_(func) {}

Logger::Logger(const std::shared_ptr<ILogger> logger, const LogFunction& func)
    : logger_(logger), func_(func) {}

void Logger::log(const char* file, int line, const char* name, int level, const char* message) {
  // func_ (LogFunction) has higher priority than logger_ (ILogger)
  if (func_) {
    func_(file, line, name, level, message, func_arg_);
  } else if (logger_) {
    logger_->log(file, line, name, level, message);
  }
}

void Logger::logger(std::shared_ptr<ILogger> logger) {
  logger_ = logger;
}

std::shared_ptr<ILogger> Logger::logger() const {
  return logger_;
}

void Logger::func(LogFunction log_func, void* arg) {
  func_ = log_func;
  func_arg_ = arg;
}

LogFunction Logger::func() const {
  return func_;
}

void* Logger::arg() const {
  return func_arg_;
}

void Logger::pattern(const char* pattern) {
  if (logger_) { logger_->pattern(pattern); }
  pattern_ = pattern;
}

const char* Logger::pattern() const {
  if (logger_) { return logger_->pattern(); }
  return pattern_.c_str();
}

void Logger::level(int level) {
  if (logger_) { logger_->level(level); }
  level_ = level;
}

int Logger::level() const {
  if (logger_) { return logger_->level(); }
  return level_;
}

void Logger::redirect(int level, void* output) {
  bool valid = false;
  if (level >= 0 && level <= kMaxSeverity) {
    // Resize the sinks_ vector if necessary
    if (static_cast<int>(sinks_.size()) <= level) { sinks_.resize(level + 1); }
    valid = true;
  }
  if (logger_) { logger_->redirect(level, output); }
  if (valid) { sinks_[level] = output; }
}

void* Logger::redirect(int level) const {
  if (logger_) { return logger_->redirect(level); }
  if (level >= 0 && level < static_cast<int>(sinks_.size())) { return sinks_[level]; }
  return nullptr;
}

}  // namespace logger

}  // namespace nvidia
